Feature/attri2vec

daokunzhang opened the pull request:
In this branch, I implemented the attri2vec model and evaluate its performance in inferring the representations of out-of-sample nodes through the out-of-sample link prediction task.

I made following three changes:

1. I added the stellargraph/demos/embeddings/stellargraph-attri2vec-DBLP.ipynb file, which shows the work flow for training attri2vec on in-sample subgraphs, and inferring the representations of out-of-sample nodes, following by the out-of-sample link prediction experiments.

2. I added the attri2vecLinkGenerator class to the stellargraph/stellargraph/mapper/link_mappers.py, which is used to generate batches of (feature of target node, id of context node, label of node pair) lists for training the attri2vec model.

3. I changed the stellargraph/demos/embeddings/README.md file, by adding the description for the attri2vec demo.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/stellargraph/stellargraph?pullRequest=470) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/stellargraph/stellargraph?pullRequest=470) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/stellargraph/stellargraph?pullRequest=470) it.</sub>

PantelisElinas commented:
Hi @daokunzhang 

please update the code to wrap attri2vec into a layer, write unit tests, and update documentation so we can review.

Regards,

P.

daokunzhang commented:
Hi @PantelisElinas ,

I have made following updates for the code:

1. Write a Attri2Vec class in file stellargraph/stellargraph/layer/attri2vec.py to wrap attri2vec in a layer, and update the stellargraph/stellargraph/layer/__init__.py file to help import Attri2Vec from the package stellargraph.layer.

2. Write a Attri2VecNodeGenerator class at file stellargraph/stellargraph/mapper/node_mappers.py to predict node representations with the learned attri2vec model.

3. Rewrite the demo file --stellargraph/demos/embeddings/stellargraph-attri2vec-DBLP.ipynb-- with the updated layers and mappers.

4. Write the unit test for Attri2Vec class at file stellargraph/tests/layer/test_attri2vec.py, the unit test for Attri2VecLinkGenerator at file stellargraph/tests/test_linkmappers.py, the unit test for Attri2VecNodeGenerator at file stellargraph/tests/test_nodemappers.py

5. Update the stellargraph/docs/api.txt file to add the index to the new added class: Attri2Vec, Attri2VecLinkGenerator, Attri2VecNodeGenerator.

Best regards!
Daokun

daokunzhang commented on stellargraph/layer/attri2vec.py:
```diff
@@ -0,0 +1,212 @@
+        if normalize == "l2":
+            self._normalization = Lambda(lambda x: K.l2_normalize(x, axis=-1))
+
+        elif normalize is None or normalize == "none" or normalize == "None":
```
For the normalize option check, I follow the code of GraphSAGE layer, to make the implementation of Attri2Vec layer consistent with GraphSAGE.

daokunzhang commented on stellargraph/layer/attri2vec.py:
```diff
@@ -0,0 +1,212 @@
+            self.input_feature_size = input_dim
+            self.input_node_num = node_num
+
+        else:
```
The Attri2Vec layer can be constructed from a generator, which only passes the parameter of the number of nodes and the dimension of features. The Attri2Vec layer can also be directly constructed with these two parameters. To train the Attri2Vec model, a generator is still needed to be passed to the keras.model.fit_generator() function.

daokunzhang commented on stellargraph/layer/attri2vec.py:
```diff
@@ -0,0 +1,212 @@
+        self.n_layers = len(layer_sizes)
+        self.bias = bias
+        self.dropout = dropout
+        self.activation = "sigmoid"  # the activation fuction used in hidden layers
```
It can be other options, like ReLU. We can make it a parameter for init to be specified.

daokunzhang commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -508,6 +514,124 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
         )
 
 
+class Attri2VecNodeGenerator:
```
To make the name more informative, shall we name it as FeatureBasedNodeGenerator?

PantelisElinas replied:
Maybe NodeFeatureGenerator would be more informative; I just swapped the words Node and Feature for your suggestion.

daokunzhang commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -508,6 +514,124 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+            raise TypeError("Graph must be a StellarGraph object.")
+
+        self.graph = G
+        self.num_samples = [0]  # for compatibility with GraphSAGE
```
The classes of LinkSequence, OnDemandLinkSequence and NodeSequence need the parameter num_samples in their init functions to set the _sampling_schema parameter for GraphSAGE. Actually the num_samples parameter is not necessary for Attri2VecLinkGenerator or Attri2VecNodeGenerator, but we need to make some changes in the LinkSequence, OnDemandLinkSequence and NodeSequence classes. Do you think it is OK?

daokunzhang commented:
Hi @PantelisElinas ,

I have finished the changes as per your comments.

The following is some explanation:

1. I keep the name of the class Attri2VecNodeGenerator unchanged. We can change it to NodeGenerator or other names when it is generalised to other cases in the future.

2. After merging the latest develop branch, the attri2vec node classification demo on DBLP with multiple workers still performs worse than single worker. I use the single worker to run attri2vec embeddings&node classification demos on both citeseer and DBLP.

3. I delete the unnecessary num_samples and sampling_schema parameters from Attri2VecNodeGenerator and Attri2VecLinkGenerator.

4. I add a function 'get_index_for_nodes' to the StellarGraphBase class to access node index for the convenience of getting the ids of context node in attri2vec.

If you need more changes, please let me know.

Best regards!
Daokun

daokunzhang commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -505,6 +537,114 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+
+        Args:
+            head_nodes: An iterable of head nodes to perform sampling on.
+            sampling_schema: The sampling schema for the model, for compatibility with GraphSAGE and HinSAGE.
```
I have deleted the sampling schema in the docstring.

daokunzhang commented on tests/layer/test_attri2vec.py:
```diff
@@ -0,0 +1,138 @@
+    assert attri2vec.input_node_num == 4
+    assert attri2vec.n_layers == 2
+    assert attri2vec.bias
+
```
Hi Pantelis,

If graphs have no features, the Attri2VecGenerator constructor will raise a Runtime Error, which has been tested in the test_attri2vec_nodemapper_constructor_no_feats() function in the test_node_mappers.py file.

daokunzhang commented:
Hi @PantelisElinas,

Please check the latest update. I have fixed the typos and added the unit test for the get_index_for_nodes function.  I also deleted the attri2vec DBLP embedding learning and node classification demos.

If you need more changes, please let me know.

Best regards!
Daokun
